COLLABORATOR_GUIDE.md

```diff
@@ -73,6 +73,112 @@ All pull requests that modify executable code should be subjected to
+
+Breaking changes are difficult primarily because they change the fundamental
+assumptions a user of Node.js has when writing their code and can cause
+existing code to stop functioning as expected -- costing developers and users
+time and energy to fix.
+
+Because breaking (semver-major) changes are permitted to land in master at any
+time, it should be *understood and expected* that at least some subset of the
+user ecosystem *may* be adversely affected *in the short term* when attempting
+to build and use Node.js directly from master. This potential instability is
+precisely why Node.js offers distinct Stable and LTS release streams that carry
+varying explicit stability guarantees. Specifically, as a general rule,
+*breaking* changes should never land in a Stable or LTS branch unless such
+changes are considered to be absolutely necessary to address critical security
+issues.
+
+When any change is landed in master, and it is determined that the such
+changes *do*, in fact, break existing code, a decision may be made to revert
+those changes either temporarily or permanently. However, the decision to
+revert or not can often be based on many complex factors that are not easily
+codified.
+
+## Deprecations
+
+Deprecation refers to the identification of APIs that should no longer be used
+and that may be removed or modified in non-backwards compatible ways in a
+future major release of Node.js.
+
+Node.js uses three fundamental Deprecation levels:
+
+* Soft-Deprecation (or Docs-only deprecation) refers to elements of the Node.js
+  public API that are being staged for deprecation in a future Node.js major
+  release. An explicit notice indicating the deprecated status is added to the
+  API documentation *but no functional changes are implemented in the code*.
+  There will be no runtime deprecation warning.
+
+* Hard-Deprecation refers to the use of warnings being reported at
+  runtime when a deprecated API is used. By default, Node.js will print a
+  warning to stderr the first time a deprecated API is used. A command-line
+  switch can be used to escalate such warnings into runtime errors that will
+  cause the Node.js process to exit. As with Soft-Deprecation, the documentation
+  for the API must be updated to clearly indicate the deprecated status.
+
+* End-of-life refers to APIs that have gone through Hard-Deprecation and are
+  ready to be removed from Node.js entirely.
+
+All Deprecations *must* be handled as breaking (semver-major) changes. This
+means that while a PR adding a Deprecation can be landed in master at any time
+with appropriate review and approval, such changes are only permitted to be
+delivered in new major releases.
+
+<a id="deprecation-cycle"></a>
+A "Deprecation cycle" is one full Node.js major release during which an API
+has been in one of the three Deprecation levels.
+
+No API can be moved to End-of-life without first having gone through a
+Hard-Deprecation cycle.
```

Comment:
What API is "no API" needs to be thoroughly discussed. For example, does this include all properties on all `_handle` objects created from one of the C++ `*Wrap` constructors? Does this include properties that are intentionally "hidden" by being placed in a `Symbol` instead of as a named property?

I don't think an acceptable solution is to say "we can change private properties without deprecation, but if we find it adversely affects the ecosystem then it will be reverted" (e.g. graceful-fs). Though it'd also make development a blood-soaked, nightmare-ish hellscape if we also say that everything exposed to the public must be deprecated.

Honestly I feel like we just need to put our foot down and say that changes to internals is not of concern; while at the same time being more responsive to users' needs and actually adding new APIs that give them the access they need, instead of having a never-ending discussion as to whether we want to support such an API forever (which IMO puts us in a worse position, as it solidifies the use of the internal property).